src/libcore/slice/mod.rs

```diff
@@ -3872,10 +3876,14 @@ unsafe impl<'a, T> TrustedRandomAccess for ExactChunksMut<'a, T> {
 /// ```
 ///
 /// [`NonNull::dangling()`]: ../../std/ptr/struct.NonNull.html#method.dangling
+/// [`pointer::offset`]: ../../std/primitive.pointer.html#method.offset
 #[inline]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub unsafe fn from_raw_parts<'a, T>(data: *const T, len: usize) -> &'a [T] {
     debug_assert!(data as usize % mem::align_of::<T>() == 0, "attempt to create unaligned slice");
+    let size = mem::size_of::<T>();
+    debug_assert!(size == 0 || len < (isize::MAX as usize + size - 1) / size,
+                  "attempt to create slice covering half the address space");
```

Comment:
Hmm, since the docs phrase the restriction in bytes, maybe write the assertion that way?  Perhaps
```rust
debug_assert!(mem::size_of::<T>().saturating_mul(len) <= isize::MAX as usize);
```
That handles ZSTs automatically and avoids non-obvious rounding code.

Reply:
I like this!